azure-iot-hub-devicesdk/azure/iot/hub/devicesdk/message_queue.py

```diff
@@ -0,0 +1,142 @@
+import queue
+import logging
+from threading import Event
+
+logger = logging.getLogger(__name__)
+
+
+class MessageQueue(queue.Queue):
+    def __init__(self, on_enable, on_disable):
+        super(MessageQueue, self).__init__()
+        self._enabled = False
+        self._on_enable = on_enable
+        self._on_disable = on_disable
+
+    def enable(self):
+        self._enabled = True
+        self._on_enable()
+
+    def disable(self):
+        self._enabled = False
+        self._on_disable()
+
+    @property
+    def enabled(self):
+        return self._enabled
+
+    @enabled.setter
+    def enabled(self, v):
+        if v:
+            self.enable()
+        else:
+            self.disable()
+
+
+class MessageQueueManager(object):
+    def __init__(self, transport):
+        self.c2d_message_queue = None
+        self.input_message_queues = {}
+        self._transport = transport
+
+    def _input_message_queue_on_enable(self):
+        self._enable_feature("input")
+        # Maybe check if input is enabled first?
+
+    def _c2d_message_queue_on_enable(self):
+        self._enable_feature("c2d")
+
+    def _input_message_queue_on_disable(self):
+        # Disable input messages as a feature if all input queues are disabled
+        if all(not queue.enabled for queue in self.input_message_queues.values()):
+            self._disable_feature("input")
+
+    def _c2d_message_queue_on_disable(self):
+        self._disable_feature("c2d")
+
+    def _enable_feature(self, feature_name):
```

Comment:
I'm considering moving these back to the client, since they'll be different depending on sync/async client, and it would be ideal to have this manager be sync/async agnostic